### PR TITLE
Key rotation Issue#140

### LINKED ISSUE #140: Handle Cosmos Key Rotation

### PR DESCRIPTION
**Pull request by sapinderpalsingh**
# Type of PR

- [ ] Documentation changes
- [X] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR
#140 
## Validation

- [ ] Unit tests updated and ran successfully
- [ ] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #140 (this will automatically close the issue when the PR closes)
- References #issue_number (this references the issue but does not close with PR)

### CODE REVIEW

File: src/app/Controllers/BaseController.cs [hunk 1]
```diff
@@ -84,5 +98,29 @@ public static JsonResult CreateResult(string message, HttpStatusCode statusCode)
...
+        private AsyncRetryPolicy GetCosmosRetryPolicy()
+        {
+            return Policy.Handle<CosmosException>(e => e.StatusCode == HttpStatusCode.Unauthorized)
+                .WaitAndRetryAsync(1, retryAttempt =>
+                 {
+                     var timeToWait = TimeSpan.FromSeconds(Math.Pow(2, retryAttempt));
+                     logger.LogInformation("Reloading configurations to check if Key was rotated.");
+                     
+                     //Reload the configurations.
+                     (configuration as IConfigurationRoot).Reload();
```
Comment (bartr) [inline src/app/Controllers/BaseController.cs]: This is an expensive operation and needs to be encapsulated so every request is not running a reload (which is n+1 KeyVault requests with n being the number of secrets in KV). 

At scale, the key vault requests will get throttled which will result in errors.

File: src/app/Controllers/ErrorResult.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using Microsoft.AspNetCore.Mvc;
```
Comment (bartr) [inline src/app/Controllers/ErrorResult.cs]: This should use the standard error handling which uses the IETF standard to return the error.

File: src/app/Controllers/BaseController.cs (hunk 1 above)
Comment (sapinderpalsingh) [reply src/app/Controllers/BaseController.cs]: @bartr,  We can refresh the connection without reloading the configurations using IKeyVaultConnection but in that case our configuration store will still be having the old key while Cosmos dal connection will be using the new Key.
So it is just the keys will be out of sync until the app restarts.
Will it have any impact or we are good with this?

**Comment by sapinderpalsingh** [comment]
@bartr , I did some code refactoring and removed the configuration reload.
Also added semaphoreSlim.WaitAsync() to prevent the race conditions and avoid keyvault throttling issues.
Currently implemented this in the Genres controller, but once it gets reviewed we can implement it for rest of the cosmos calls.

**Review by bartr — APPROVED** [review]

**Pull request merged**